File: C++/problem2_2_1.cpp
/*
*
* @file problem2_2_1.cpp
* @author dengjiaying (511492494@qq.com)
* @brief
* @version 0.1
* @date 2022-04-12
*
* @copyend Copyend (c) 2022
*
*/
#include<opencv2/opencv.hpp>
#include<iostream>
#include<time.h>
#include<string>
#include<vector>
using namespace std;
using namespace cv;

void Rgb2Hsv(float R, float G, float B, float& H, float& S, float&V)
{
    // r,g,b的值是0到1之间的实数
    // h = [0,360], s = [0,1], v = [0,1]
    // if s == 0, then h = -1 (undefined)
    float min, max, delta, tmp;
    //三目运算符 给min赋值（min为r,g,b中的最小值）
    tmp = R>G ? G : R;
    min = tmp>B ? B : tmp;
    //三目运算符 给max赋值（min为r,g,b中的最大值）
    tmp = R>G ? R : G;
    max = tmp>B ? tmp : B;
    //s（饱和度）  v(亮度)
    V = max / 255 * 100; // v
    //计算公式中的分母
    delta = max - min;

    if (max != 0)
        S = delta / max *100; // s
    else
    {
        // r = g = b = 0 // s = 0, v is undefined
        S = 0;
        H = 0;
        return;
    }
    //灰色
    if (delta == 0)
    {
        H = 0;
        return;
    }
    //s，v计算公式
    else if (R == max)
    {
        if (G >= B)
            H = (G - B) / delta; //yellow 到 magenta
        else
            H = (G - B) / delta + 6.0;
    }
    else if (G == max)
        H = 2.0 + (B - R) / delta; //cyan 到 yellow
    else if (B == max)
        H = 4.0 + (R - G) / delta; //magenta 到 cyan

    H *= 60.0; // degrees（计算公式）

}
int main()
{
    Mat srcImage =cv::imread("E:\\lena.jpg");
    //如果图片不存在
    if(srcImage.empty())
    {
        cout << "can't read this image!" << endl;
        return -1;
    }

    Mat gray = Mat::zeros(srcImage.rows, srcImage.cols, CV_8UC1);
    for (int i = 0; i < srcImage.rows; i++)
    {
        for (int j = 0; j < srcImage.cols; j++)
        {
            //查找两数最大值并返回最大数
            gray.at<uchar>(i, j) = fmax(srcImage.at<Vec3b>(i, j)[0],fmax(srcImage.at<Vec3b>(i, j)[1], srcImage.at<Vec3b>(i, j)[2]));
        }
    }

    namedWindow("img", WINDOW_AUTOSIZE);
    namedWindow("gray", WINDOW_AUTOSIZE);
    //展示原图像
    imshow("img",srcImage);
    //展示灰度图
    imshow("gray", gray);

    int row = srcImage.rows;
    int col = srcImage.cols;
    //创建HSV图像
    Mat hsvimg= Mat(row, col ,CV_8UC3);

    for (int i = 0; i < row; i++)
    {
        for (int j = 0; j < col; j++)
        {
            float b = srcImage.at<Vec3b>(i, j)[0];
            float g = srcImage.at<Vec3b>(i, j)[1];
            float r = srcImage.at<Vec3b>(i, j)[2];
            float h, s, v;

            Rgb2Hsv(r, g, b, h, s, v);
            hsvimg.at<Vec3b>(i, j)[0] = h;//h数值若大于255会自动减去255
            hsvimg.at<Vec3b>(i, j)[1] = s;
            hsvimg.at<Vec3b>(i, j)[2] = v;
        }
    }
    //展示HSV图像
    imshow("hsv",hsvimg);

    waitKey(0);
    return 0;
}

File: Python/problem2_2_1.py
# !/usr/bin/env python
# -*- coding: utf-8 -*-
# @Time    :# ${2022/4/12}
# @Author  : dengjiaying
# @File    : $problem2_2_1.py
# @Brief   :
# @Version : 0.1
import numpy as np
import cv2 as cv

#def自定义函数框架（RGB转灰度图）
def Gray_img(img):
    print(img.shape)
    (h, w, c) = img.shape
    #图像单个通道的像素：
    img_b = img[:, :, 0]
    img_g = img[:, :, 1]
    img_r = img[:, :, 2]
    #通过公式G=0.114*B+0.587*G+0.229*R来实现灰度化：
    img_gray = img_r * 0.299 + img_g * 0.587 + img_b * 0.114
    # 用于表示图像的numpy数组数据类型必须为整型，否则无法输出结果
    img_gray = img_gray.astype(np.uint8)
    #展示原图像：
    cv.imshow('img',img)
    #展示灰度图：
    cv.imshow('img_gray', img_gray)
    cv.waitKey(0)
    return 0
#def自定义函数框架（RGB转HSV）
def rgb2hsv(img):
    # 图像的高度
    h = img.shape[0]
    # 图像的宽度
    w = img.shape[1]
    # h行w列的矩阵：
    H = np.zeros((h,w),np.float32)
    S = np.zeros((h, w), np.float32)
    V = np.zeros((h, w), np.float32)
    # 图像通道的分离:
    r,g,b = cv.split(img)
    # 对图像进行归一化，范围为[0,1]:
    r, g, b = r/255.0, g/255.0, b/255.0
    #h,s,v计算公式
    for i in range(0, h):
        for j in range(0, w):
            mx = max((b[i, j], g[i, j], r[i, j]))
            mn = min((b[i, j], g[i, j], r[i, j]))
            V[i, j] = mx
            if V[i, j] == 0:
                S[i, j] = 0
            else:
                S[i, j] = (V[i, j] - mn) / V[i, j]
            if mx == mn:
                H[i, j] = 0
            elif V[i, j] == r[i, j]:
                if g[i, j] >= b[i, j]:
                    H[i, j] = (60 * ((g[i, j]) - b[i, j]) / (V[i, j] - mn))
                else:
                    H[i, j] = (60 * ((g[i, j]) - b[i, j]) / (V[i, j] - mn))+360
            elif V[i, j] == g[i, j]:
                H[i, j] = 60 * ((b[i, j]) - r[i, j]) / (V[i, j] - mn) + 120
            elif V[i, j] == b[i, j]:
                H[i, j] = 60 * ((r[i, j]) - g[i, j]) / (V[i, j] - mn) + 240
            H[i,j] = H[i,j] / 2
    return H, S, V
#读取图像
img = cv.imread('E:\\lena.jpg')

h,s,v = rgb2hsv(img)

img_gray = Gray_img(img)
#展示三种通道的图片：
cv.imshow("h",h)
cv.imshow("s",s)
cv.imshow("v",v)
#前面分离出来的三个通道
merged = cv.merge([h,s,v])
#展示HSV图片：
cv.imshow("hsv",merged)
cv.waitKey(0)
cv.destroyAllWindows()

